sky/serve/spot_policy.py

```diff
@@ -0,0 +1,132 @@
+""" Sky Spot Policy for SkyServe."""
+import enum
+import random
+import typing
+from typing import Dict, List, Optional, Type
+
+from sky import sky_logging
+
+if typing.TYPE_CHECKING:
+    from sky.serve import replica_managers
+    from sky.serve import service_spec
+
+logger = sky_logging.init_logger(__name__)
+
+
+class SpotZoneType(enum.Enum):
+    """Spot Zone Type."""
+    ACTIVE = 'ACTIVE'
+    PREEMPTED = 'PREEMPTED'
+
+
+class SpotPlacer:
+    """Spot Placement specification."""
+    NAME: Optional[str] = None
+    REGISTRY: Dict[str, Type['SpotPlacer']] = dict()
+
+    def __init__(self, spec: 'service_spec.SkyServiceSpec') -> None:
+        assert spec.spot_zones is not None
+        self.zones = list(spec.spot_zones)
+        self.zone2type: Dict[str, SpotZoneType] = {
+            zone: SpotZoneType.ACTIVE for zone in self.zones
+        }
+
+    def __init_subclass__(cls) -> None:
+        if cls.NAME is None:
+            # This is an abstract class, don't put it in the registry.
+            return
+        assert cls.NAME not in cls.REGISTRY, f'Name {cls.NAME} already exists'
+        cls.REGISTRY[cls.NAME] = cls
+
+    @classmethod
+    def get_policy_names(cls) -> List[str]:
+        return list(cls.REGISTRY.keys())
+
+    def select(self, existing_replicas: List['replica_managers.ReplicaInfo'],
+               num_zones: int) -> List[str]:
+        """Select next zone to place spot instance."""
+        raise NotImplementedError
```

Comment:
nit: variable name `num_zones` might not express the meaning correctly, considering only 1 zone is enabled and you want to make the decision for 2 replicas. It will only return 1 zone but with 2 entries in the list.

Reply:
changed it to `num_replicas`